### PR TITLE
Introduce a minimum CGU size in non-incremental builds.

### PR DESCRIPTION
**Pull request by nnethercote**
Because tiny CGUs slow down compilation *and* result in worse generated code.

r? @wesleywiser

### CODE REVIEW

**Comment by nnethercote** [comment]
I got some `tests/run-make` failures locally. I've now fixed them, but I'll wait for the perf run to complete before uploading the relevant changes.

**Comment by nikic** [comment]
@nnethercote Could the fact that the CGUs now get sorted by size even if there are less than the limit make a difference?

**Comment by nnethercote** [comment]
> @nnethercote Could the fact that the CGUs now get sorted by size even if there are less than the limit make a difference?

I had a similar thought, but [this](https://github.com/rust-lang/rust/pull/112448/files#diff-abf5c721a877d6f5909d0f8dbe78f085adcc67fab0b01b90b7bb83fb68f672daL404) runs shortly afterwards, sorting the CGUs by name. Also, I did some debug logging and the CGU contents were identical before and after. Even worse, I'm getting big discrepancies between instruction counts measurements from Cachegrind vs perf. It's all very puzzling, I'll have to investigate more next week.

**Comment by nikic** [comment]
Could this be because the compiler-builtins no longer respects the `codegen-units=10000` option? Probably that one is subject to a lot of merging now.

(My first thought was that it's related to libstd codegen units, but we build with `codegen-units-std=1` in CI.)

**Comment by nnethercote** [comment]
> Could this be because compiler-builtins no longer respects the `codegen-units=10000` option? Probably that one is subject to a lot of merging now.
 
Oh, that's an interesting thought, thank you. I have just pushed new code with a temporary hack that disables the too-small merging if the cgu count is exactly 10,000, to test this theory directly.

@bors try @rust-timer queue

BTW, I worked out that you were right earlier, and that the old version of the code did cause the CGUs to end up in a different order. But the most recent perf run was on an updated version where I avoided that, and the perf results were no different. So in practice the different order didn't have a perf effect.

**Comment by nnethercote** [comment]
> I have just pushed new code with a temporary hack that disables the too-small merging if the cgu count is exactly 10,000, to test this theory directly.

Good news! Those results are substantially better. Thanks again, @nikic.

There are still some oddities.
- We are still getting some perf changes for incremental builds that should be completely unaffected by this PR. Some of them could be noise. But binary size is very reliable, and we got binary size reductions for incremental builds of `ripgrep` and `tt-muncher`, which is really weird.
- Every metric has great results, *except* for walltime, which is also the most important metric. The results there are a real mix, including some unexpected incremental changes. Overall the walltime wins and losses roughly balance out. With all the other metric benefits -- especially max-rss reductions, binary size reductions, and less parallelism (which is good for reducing jobserver competition when multiple rustc processes are running at the same time) -- this will definitely be worth merging even if it's not a walltime win.

Before I do that, I have a couple of things to do:
- Remove the 10,000 CGU hack. I plan to distinguish between the default CGU count values vs. those provided by a user. When the user specifies a count, I will not do the merging of tiny CGUs. As well as solving the 10,000 case, this will be useful in testing, because we have several tests that use a specific CGU count (e.g. 6) and expect the produced code to have a very particular form.
- Try some values other than 1000 for the minimum size.

**Comment by nnethercote** [comment]
I removed the 10,000 hack and changed the minimum size to 500.

@bors try @rust-timer queue

**Comment by nikic** [comment]
Looks like the bot did not pick up the try build.

@rust-timer build d1abee25f15fa534d479113260ad8337fb156835

**Comment by nnethercote** [comment]
500 gives worse results than 1000. Let's try 2000.

@bors try @rust-timer queue

File: compiler/rustc_session/src/session.rs [hunk 1]
```diff
@@ -234,6 +234,27 @@ pub enum MetadataKind {
...
+    User(usize),
+
+    /// A default value, i.e. not specified by the user. In this case we take
+    /// more liberties about CGU formation, e.g. avoid producing very small
+    /// CGUs.
+    Default(usize),
+}
+
+impl CodegenUnits {
+    pub fn inner(self) -> usize {
```
Comment (wesleywiser) [inline compiler/rustc_session/src/session.rs:250]: nit: I wonder if naming this `as_usize` would make the callsites read better?

File: src/doc/rustc/src/codegen-options/index.md [hunk 2]
```diff
@@ -31,8 +31,9 @@ Supported values can also be discovered by running `rustc --print code-models`.
 
 ## codegen-units
 
-This flag controls how many code generation units the crate is split into. It
-takes an integer greater than 0.
+This flag controls the maximum number of code generation units the crate is
+split into. (The exact number will depend on the size and structure of the
+crate's source code.) It takes an integer greater than 0.
```
Comment (wesleywiser) [inline src/doc/rustc/src/codegen-options/index.md:36]: I could see the parenthetical statement being interpreted as "we may create more CGUs than the value you specified". Maybe this formulation is more clear on that point? 
 
```suggestion
split into. (Fewer code generation units may be created depending on the 
size and structure of the crate's source code.) It takes an integer 
greater than 0.
```

**Review by wesleywiser — APPROVED** [review]

File: compiler/rustc_session/src/session.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_session/src/session.rs:250]: Good idea.

File: src/doc/rustc/src/codegen-options/index.md (hunk 2 above)
Comment (nnethercote) [reply src/doc/rustc/src/codegen-options/index.md:36]: I'll remove the parenthetical. The important part of the change is specifying that it's a maximum.

**Comment by nnethercote** [comment]
> My guess is that this is due to fewer duplicated inline functions in different CGUs. This might be another statistic worth reporting on.

Yes, that's definitely a big part of the improvement here. It's hard to report on because merging currently happens before inlining, so you can't make this measurement without two versions of the compiler, one doing the tiny CGU merging and one not.

**Comment by nnethercote** [comment]
I have addressed the review comments. I also made some very small tweaks to the debug printing, but nothing needing additional review. Thanks!

@bors r=wesleywiser

**Pull request merged**